### PR TITLE
Improve entry attachments view

### PR DESCRIPTION
**Pull request by frostasm**
Improve entry attachments view.

## Description

- Create separate widget to edit/view entry attachments - **EntryAttachmentsWidget**
- Add drag and drop support for attachments view in EntryAttachmentsWidget
- Use EntryAttachmentsWidget to display and edit  attachments in EditEntryWidget
- Add attachments tab to details view - use EntryAttachmentsWidget
- Allow to open attachments from details view

## Motivation and context

In my opinion, the tableview more user friendly than a listview.

## How has this been tested?

Manually.

## Screenshots (if appropriate):

**EditEntryWidget** 

image

**Details view**

image

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/gui/DetailsWidget.ui [hunk 1]
```diff
@@ -130,7 +130,7 @@
    <item>
     <widget class="QTabWidget" name="tabWidget">
      <property name="currentIndex">
-      <number>0</number>
+      <number>4</number>
```
Comment (TheZ3ro) [inline src/gui/DetailsWidget.ui]: Revert this to 0. QtCreator mixed this up

Comment (frostasm) [reply src/gui/DetailsWidget.ui]: Maybe it is better to activate the first tab in the constructor (to not depend on the form state)?

File: tests/TestEntryModel.cpp [hunk 2]
```diff
@@ -122,9 +122,11 @@ void TestEntryModel::testAttachmentsModel()
     entryAttachments->set("first", QByteArray("123"));
 
     entryAttachments->set("2nd", QByteArray("456"));
-    entryAttachments->set("2nd", QByteArray("789"));
+    entryAttachments->set("2nd", QByteArray("7895"));
```
Comment (TheZ3ro) [inline tests/TestEntryModel.cpp]: Why this change?

File: src/gui/DetailsWidget.cpp [hunk 3]
```diff
@@ -338,3 +359,28 @@ void DetailsWidget::updateTabIndex(int index)
         m_selectedTabEntry = index;
     }
 }
+
+void DetailsWidget::openAttachment(const QModelIndex& index)
```
Comment (TheZ3ro) [inline src/gui/DetailsWidget.cpp]: Can't we use the `openAttachment` from the EditEntryWidget class? Or alternatively declaring it static in a different class and point both of them to the static one? This is just repeated code and I would like to leave the detailsWidget light

File: src/gui/DetailsWidget.ui (hunk 1 above)
Comment (TheZ3ro) [reply src/gui/DetailsWidget.ui]: I think the currentIndex is needed in the xml anyway, just set it to `0` so we don't add extra code for it

File: src/gui/DetailsWidget.cpp (hunk 3 above)
Comment (frostasm) [reply src/gui/DetailsWidget.cpp]: Agree with you. I'll fix it tonight.

File: tests/TestEntryModel.cpp (hunk 2 above)
Comment (frostasm) [reply tests/TestEntryModel.cpp]: Сhanged to have different size ("456" - 3 bytes, "7895" - 4 bytes,  I check the size in the next lines). This is not critical, I can remove the changes.

File: src/gui/DetailsWidget.cpp (hunk 3 above)
Comment (frostasm) [reply src/gui/DetailsWidget.cpp]: @TheZ3ro perhaps it is better to create a separate widget to display/edit the attachments. With the functionality of adding, deleting and opening files (which can be turned off in the details view).

Comment (TheZ3ro) [reply src/gui/DetailsWidget.cpp]: It's fine opening files from the details view, it's the extra code that I don't really like.
Making a separate widget and import it will be very nice and convenient

File: tests/TestEntryModel.cpp (hunk 2 above)
Comment (TheZ3ro) [reply tests/TestEntryModel.cpp]: Just to be consistent with the `123-456-789` can you replace the final `5` with a `0` ? 
Merely style reasons

**Comment by frostasm** [comment]
@TheZ3ro I added a widget to edit/view the attachments

**Comment by frostasm** [comment]
@TheZ3ro I rebase changes on top of the develop branch ;)

**Comment by fonic** [comment]
@frostasm this is great, good work!

**Comment by TheZ3ro** [comment]
This is working fine and ready to merge, waiting someone else review, ping @keepassxreboot/core-developers 

Nice PR follow-up idea: add keyboard actions for enter and cancel keys for open and remove attachments

**Comment by frostasm** [comment]
@fonic thank you :)


> Nice PR follow-up idea: add keyboard actions for enter and cancel keys for open and remove attachments

@TheZ3ro good idea. What keyboard shortcuts would be helpful for each role?
- **add** - insert key
- **remove** - delete key
- **open** - [QAbstractItemView::activated() signal](http://doc.qt.io/qt-5/qabstractitemview.html#activated) 
- **save** - ?

Possible improvements:

- ask permission to replace the existing attachments, when adding new attachments
- add the ability to sort attachments (by name, by size)
- add the ability to rename existing attachments


I have free time on the weekends and can implement the described proposal.

**Comment by TheZ3ro** [comment]
@frostasm can you report those proposal in a new issue? So we can track them for a new PR. 
This PR is fine as-is

**Comment by frostasm** [comment]
@TheZ3ro okay 👍

**Comment by frostasm** [comment]
@droidmonkey can you review the changes?

**Comment by droidmonkey** [comment]
Probably not for a while

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+}
+
+bool EntryAttachmentsWidget::openAttachment(const QModelIndex& index, QString* errorMessage)
+{
+    const QString filename = m_attachmentsModel->keyByIndex(index);
+    const QByteArray attachmentData = m_entryAttachments->value(filename);
+
+    // tmp file will be removed once the database (or the application) has been closed
+    const QString tmpFileTemplate = QDir::temp().absoluteFilePath(QString("XXXXXX.").append(filename));
+    QTemporaryFile* tmpFile = new QTemporaryFile(tmpFileTemplate, this);
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: Use QScopedPointer and remove following deletes.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    if (!saveDir.exists()) {
+        if (saveDir.mkpath(saveDir.absolutePath())) {
+            errorOccurred(tr("Unable to create the directory:\n").append(saveDir.absolutePath()));
+            return;
+        }
+    }
+    config()->set("LastAttachmentDir", QFileInfo(saveDir.absolutePath()).absolutePath());
+
+    QStringList errors;
+    for (const QModelIndex &index: indexes) {
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: & belongs to type.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+void EntryAttachmentsWidget::openSelectedAttachments()
+{
+    const QModelIndexList indexes = m_ui->attachmentsView->selectionModel()->selectedRows(0);
+    if (indexes.isEmpty()) {
+        return;
+    }
+
+    QStringList errors;
+    for (const QModelIndex &index: indexes) {
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: & belongs to type.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+    const bool hasSelection = m_ui->attachmentsView->selectionModel()->hasSelection();
+
+    m_ui->addAttachmentButton->setEnabled(!m_readOnly);
+    m_ui->removeAttachmentButton->setEnabled(hasSelection && !m_readOnly);
+
+    m_ui->saveAttachmentButton->setEnabled(hasSelection);
+    m_ui->openAttachmentButton->setEnabled(hasSelection);
+}
+
+bool EntryAttachmentsWidget::insertAttachments(const QStringList& filenames, QString* errorMessage)
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: Why pointer to QString instead of non-const reference?

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+            }
+            break;
+        }
+        case QEvent::Drop: {
+            QDropEvent* dropEv = static_cast<QDropEvent*>(e);
+            const QMimeData* mimeData = dropEv->mimeData();
+            if (mimeData->hasUrls()) {
+                dropEv->acceptProposedAction();
+                QStringList filenames;
+                for (const QUrl url: mimeData->urls()) {
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: Use asConst() around mimeData->urls() to avoid detaching.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+                QString errorMessage;
+                if (!insertAttachments(filenames, &errorMessage)) {
+                    errorOccurred(errorMessage);
+                }
+
+                return true;
+            }
+        }
+        default:
+            break;
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: Statement does nothing, remove default clause.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+                }
+
+                QString errorMessage;
+                if (!insertAttachments(filenames, &errorMessage)) {
+                    errorOccurred(errorMessage);
+                }
+
+                return true;
+            }
+        }
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: Add break here to avoid bugs if someone adds a new clause.

File: src/gui/entry/EntryAttachmentsWidget.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    void openSelectedAttachments();
+    void updateButtonsEnabled();
+
+private:
+    bool insertAttachments(const QStringList& fileNames, QString* errorMessage);
+    bool openAttachment(const QModelIndex& index, QString* errorMessage);
+
+    bool eventFilter(QObject* watched, QEvent* event) override;
+
+    Ui::EntryAttachmentsWidget* m_ui;
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.h]: Use QScopedPointer and remove delete from destructor.

File: src/gui/entry/EntryAttachmentsWidget.h [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    EntryAttachments* m_entryAttachments;
+    EntryAttachmentsModel* const m_attachmentsModel;
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.h]: These two should be QPointers.

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    if (errorMessage && !errors.isEmpty()) {
+        *errorMessage = tr("Unable to open files:\n%1").arg(errors.join('\n'));
+    }
+
+    return errors.isEmpty();
+}
+
+bool EntryAttachmentsWidget::openAttachment(const QModelIndex& index, QString* errorMessage)
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.cpp]: non-const reference for errorMessage instead of pointer to QString.

File: src/gui/entry/EditEntryWidgetAdvanced.ui [hunk 14]
```diff
@@ -6,8 +6,8 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>400</width>
-    <height>366</height>
+    <width>448</width>
+    <height>370</height>
```
Comment (phoerious) [inline src/gui/entry/EditEntryWidgetAdvanced.ui]: You should completely remove widget sizes here (click the little reset button in QtDesigner's property view).

File: src/gui/entry/EntryAttachmentsWidget.ui [hunk 15]
```diff
@@ -0,0 +1,107 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
+ <class>EntryAttachmentsWidget</class>
+ <widget class="QWidget" name="EntryAttachmentsWidget">
+  <property name="geometry">
+   <rect>
+    <x>0</x>
+    <y>0</y>
+    <width>589</width>
+    <height>300</height>
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsWidget.ui]: Remove widget sizes.

File: src/gui/DetailsWidget.cpp [hunk 16]
```diff
@@ -75,9 +87,10 @@ void DetailsWidget::getSelectedEntry(Entry* selectedEntry)
 
     m_ui->stackedWidget->setCurrentIndex(EntryPreview);
 
-    if (m_ui->tabWidget->count() < 4) {
-        m_ui->tabWidget->insertTab(static_cast<int>(AttributesTab), m_attributesWidget, "Attributes");
-        m_ui->tabWidget->insertTab(static_cast<int>(AutotypeTab), m_autotypeWidget, "Autotype");
+    if (m_ui->tabWidget->count() < 5) {
```
[the remark this answers is not in the document]
Comment (frostasm) [reply src/gui/DetailsWidget.cpp]: I completely agree with you about the magic numbers. I think it is necessary to refactor DetailsWidget in another pull request. I can do that after this changes will be merged. If you don't mind.

[the remark this answers is not in the document]
Comment (phoerious) [reply src/gui/DetailsWidget.cpp]: I think it would be enough to simply clear out the widget and add stuff as needed (I think it's unlikely that this would impact performance massively).

File: src/gui/entry/EntryAttachmentsWidget.cpp (hunk 9 above)
Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: if I delete **default** statement then the compiler will be very dissatisfied)

image

File: src/gui/entry/EntryAttachmentsWidget.h (hunk 12 above)
Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.h]: I think that it is not necessary, they are only used internally (they are not re-assigned in the process of work)

File: src/gui/entry/EntryAttachmentsWidget.cpp (hunk 7 above)
Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: In my opinion, this implementation is more flexible and more expressive.
Pros:
- the caller obviously has to pass a reference that explicitly indicates that the passed argument can be changed
- in case we do not need a error message, we just send nullptr 

For example:

```cpp
func1(param, result);
// vs 
func2(param, &result);
```
In case with func2, we do not need to look at the function declaration to understand that the second parameter can be changed in the function

File: src/gui/entry/EntryAttachmentsWidget.cpp [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+    m_ui->addAttachmentButton->setEnabled(!m_readOnly);
+    m_ui->removeAttachmentButton->setEnabled(hasSelection && !m_readOnly);
+
+    m_ui->saveAttachmentButton->setEnabled(hasSelection);
+    m_ui->openAttachmentButton->setEnabled(hasSelection);
+}
+
+bool EntryAttachmentsWidget::insertAttachments(const QStringList& filenames, QString* errorMessage)
+{
+    Q_ASSERT(!isReadOnly());
```
[the remark this answers is not in the document]
Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: done

**Comment by frostasm** [comment]
@phoerious thanks for the detailed review 👍. About magic numbers and bad function names. If you don't mind, I would refactor this widget in my next PR. Because current pull request relate only to the introduction of the new functionality and not to the refactoring of the existing one.

File: src/gui/entry/EntryAttachmentsWidget.cpp (hunk 7 above)
Comment (phoerious) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: A non-const reference is already an explicit indication that the value will be passed back. In case of no error message, it just remains unchanged.
If you use a pointer, you need to explicitly check for nullptr before you dereference it inside the function, which is quite unnecessary. Passing in a pointer also requires you to allocate a string on the freestore for no apparent reason (unless you really want to go down the rabbit hole of passing pointers to stack variables to the function).

File: src/gui/entry/EntryAttachmentsWidget.h (hunk 12 above)
Comment (phoerious) [reply src/gui/entry/EntryAttachmentsWidget.h]: No, but it also doesn't hurt anybody and may catch errors further down the road. Who knows where these objects are used in 5 months...

Comment (phoerious) [reply src/gui/entry/EntryAttachmentsWidget.h]: BTW I really only mean QPointers, not QSharedPointers or QScopedPointers. QPointers simply guard the pointer and reset to 0 if the QObject is deallocated. They don't count references or own anything and they can be converted to and from raw pointers.

**Comment by phoerious** [comment]
Ok, fair enough. Just don't forget about it.

File: src/gui/entry/EntryAttachmentsWidget.cpp (hunk 9 above)
Comment (phoerious) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: Ouch. Maybe use an if/else if construct? You are only handling three cases of many and this empty default clause is extremely ugly.

Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.cpp]: Done

File: src/gui/entry/EntryAttachmentsWidget.h (hunk 12 above)
Comment (frostasm) [reply src/gui/entry/EntryAttachmentsWidget.h]: Done

**Comment by frostasm** [comment]
@phoerious I think I finished :D

File: src/gui/entry/EntryAttachmentsModel.cpp [hunk 18]
```diff
@@ -65,24 +67,48 @@ int EntryAttachmentsModel::columnCount(const QModelIndex& parent) const
...
+QVariant EntryAttachmentsModel::data(const QModelIndex& index, int role) const
+{
+    if (!index.isValid()) {
         return QVariant();
     }
+    const int column = index.column();
+    if (role == Qt::DisplayRole) {
+        const QString key = keyByIndex(index);
+        switch (column) {
+        case Columns::NameColumn:
```
Comment (phoerious) [inline src/gui/entry/EntryAttachmentsModel.cpp]: I would replace the inner switch/case statements with if/else if as well. Perhaps even flatten this structure to a single level, since DisplayRole && NameColumn have the same action as EditRole && NameColumn, so it can be reduced to a single condition. DisplayRole && SizeColumn and EditRole && SizeColumn are also very similar, so they could be reduced to a return guard and a another return if the previous condition wasn't met.

Comment (frostasm) [reply src/gui/entry/EntryAttachmentsModel.cpp]: Done ;)

Comment (phoerious) [reply src/gui/entry/EntryAttachmentsModel.cpp]: Much nicer and shorter. Thanks.

**Comment by frostasm** [comment]
@phoerious @TheZ3ro thanks :)

**Comment by phoerious** [comment]
Well, thank _you_. 😄
